### PR TITLE
Reworks block chance into Parrying.

### PR DESCRIPTION
**Pull request by hal9000PR**
## What Does This PR Do
Adds a parrying component.
no block chance has actually been converted yet. I need to settle on numbers.

now, instead of blocking with an item, you will **resist** while holding an item that can block, this will raise your guard. 

your guard will be raised for a duration. if you get hit during this duration, you will parry the attack.

when you parry you will take a small amount of stamina damage, the amount of stamina damage is relative to time since you parried, the quality of the blocking item and the damage the item would have dealt and its armour pen.

I also decreased the resist cooldown as I dont see a reason why resisting should prevent you from acting for 2 full seconds. 
I might make it use its own unique button instead of resist.

## Why It's Good For The Game
Shields have been in a poor state for a long time.

they take zero skill to use and effectively double your health (on average), something about them needs changing, and this will be a fun experiment to see how people perform with it. 

Making shields take skill to use means that you are rewarded for being good at using a shield. but you can also be heavily punished for using it poorly. 

## Testing
threw it on the mentorbox for people to break.

## Changelog
:cl:
add: parrying, press the resist key while holding an item that once had block chance to parry. converts damage into stamina damage based on how well you timed the parry.
/:cl:

### CODE REVIEW

**Comment by Leanfrog** [comment]
So, is this _just_ shields? Changing them is definitely a step in the right direction but I wonder how other passive block chance items will be handled, for example the various chaplain weapons and eswords, in particular the eswords as they already have an in hand use in the form of turning them on or off.

**Comment by trololiver112** [comment]
nice

File: code/game/objects/items/weapons/shields.dm [hunk 1]
```diff
@@ -69,6 +78,9 @@
 	attack_verb = list("shoved", "bashed")
 	var/active = FALSE
 
+/obj/item/shield/riot/buckler/add_parry_component()
+	return
+
```
Comment (Qwertytoforty) [inline code/game/objects/items/weapons/shields.dm:83]: Did you mean /obj/item/shield/energy/add_parry_component() ?

**Comment by Qwertytoforty** [comment]
cursed katana / pyro gloves do not block. 
Cursed should block only melee, pyro should block all.

**Comment by AzuleUtama** [comment]
From testing on mentorbox:

- Already ID'd as a bug, but it blocks most magic attacks which it really shouldn't do.
- Some kind of sound or visual cue for when you activate the parry or drop your parry after attacking would be nice, but not a priority.
- Armor penetration needs to be factored in, we discussed either flat stamina damage bonus or leaking damage through the block.
- Cooldown between resists is also needed, as you can spam and block literally everything thrown at you - should be rewarded for timing it well instead.
- Should it be less/ineffective if you're crawling as well?
- Should projectiles with an 'explosive' effect (BSG, rocket, frag-12 etc.) bypass the block entirely?

**Review by SteelSlayer — COMMENTED** [review]
Everything looks fine, but the use of `GetComponent` is what I'd like to avoid if possible, and so far I've failed to come up with a solution to replace it's use with signals. I'm aware that we have a lot of components that use it but it's not good practice, since signals should be the only way things interact with components (the guide to signals/components goes into this into more detail).

**Comment by AzuleUtama** [comment]
From more testing:

- Spamming is less effective than before.
- Not sure if it's intended, but all mobs can see if you're parrying with the little animation that plays on your character.
- Seems to no longer block all magic projectiles now (wands, staffs, meat hook etc.).
- Cursed katana still doesn't block anything.
- Projectiles that pass through mobs and structures (xray, penetrator rounds etc.) can be blocked - but keep going through the mob despite doing no damage.
- There is a massive difference now with anything that's armor piercing. You can block a carp user all day long, but easily go down with about 4-5 AP stechkin rounds.
- Anything that has an "effect on hit" such as immolator beams will prevent the effect so long as it's blocked.
- Cling tentacle can be blocked regardless of the intent when fired.
- Rockets cannot be blocked.
- Ranged holoparasite/guardians crystal spray attack cannot be blocked at all - regular melee can.
- Pulse rifle can easily be blocked without much stamina loss, probably due to lack of AP.

**Comment by Somniworld** [comment]
I've unironically been looking forward to something like for a while now, super glad someone more talented than I is working on it!

When you mean "resist", will it mean the *resist* button that is currently in the game (and which I do not remember having a cooldown, hence my confusion)? If that is so, how will it handle being buckled / resisting out of a grab, will it prevent you from getting out of a chair fast, or prevent from blocking while doing that?

Either way, looking forward to this being added!

**Comment by LordMadness** [comment]
Will the Captain's Rapier, Energy Sword and Double Energy Sword (etc) melee blocks also be turned into Parry?
Because if so, it's a pretty rough nerf.

I get the point that the current block takes no skill, but the Desword is 16TC item.
With the passive buff gone, which, like it or not is a big reason for its cost- What can we expect for these items?
It'd be rather cool if the cost went down and maybe add the energy shield to the Uplink?

**Comment by hal9000PR** [comment]
> What can we expect for these items?

I would be happy with a follow up pr that tweaks the values of the esword.

**Comment by hal9000PR** [comment]
after some feedback I have increased the cooldown and reduced the parry window. to help prevent people just spamming it.

**Review by SteelSlayer — APPROVED** [review]
GetComponent shall live, for now

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 2]
```diff
@@ -243,7 +233,18 @@ emp_act
...
 	return FALSE
 
-#undef BLOCK_CHANCE_CALCULATION
+/mob/living/carbon/human/proc/get_best_shield()
+	var/datum/component/parry/left_hand_parry = l_hand?.GetComponent(/datum/component/parry)
+	var/datum/component/parry/right_hand_parry = r_hand?.GetComponent(/datum/component/parry)
+	if(!right_hand_parry && !left_hand_parry)
+		return null // no parry component
+
+	if((right_hand_parry && left_hand_parry))
```
Comment (farie82) [inline code/modules/mob/living/carbon/human/human_defense.dm:242]:
```suggestion
	if(right_hand_parry && left_hand_parry)
```

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 3]
```diff
@@ -243,7 +233,18 @@ emp_act
...
+		if(right_hand_parry.stamina_coefficient > left_hand_parry.stamina_coefficient) // try and parry with your best item
+			return left_hand_parry.parent
+		else
+			return right_hand_parry.parent
+	return (right_hand_parry && right_hand_parry.parent) || (left_hand_parry && left_hand_parry.parent) // parry with whichever hand has an item that can parry
```
Comment (farie82) [inline code/modules/mob/living/carbon/human/human_defense.dm:247]:
```suggestion
	return right_hand_parry?.parent || left_hand_parry?.parent // parry with whichever hand has an item that can parry
```

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 4]
```diff
@@ -215,21 +215,11 @@ emp_act
...
 
-	if(l_hand)
-		l_block_chance = BLOCK_CHANCE_CALCULATION(l_hand.block_chance, armour_penetration_flat, armour_penetration_percentage, block_chance_modifier)
-	if(r_hand)
-		r_block_chance = BLOCK_CHANCE_CALCULATION(r_hand.block_chance, armour_penetration_flat, armour_penetration_percentage, block_chance_modifier)
-
-	if(l_block_chance > r_block_chance && l_hand.hit_reaction(src, AM, attack_text, l_block_chance, damage, attack_type))
-		return TRUE
-	else if(r_block_chance > 0 && r_hand.hit_reaction(src, AM, attack_text, r_block_chance, damage, attack_type))
+	if(shield && shield.hit_reaction(src, AM, attack_text, 0, damage, attack_type))
```
Comment (farie82) [inline code/modules/mob/living/carbon/human/human_defense.dm:222]:
```suggestion
	if(shield?.hit_reaction(src, AM, attack_text, 0, damage, attack_type))
```

**Review by farie82 — APPROVED** [review]
Minor styling things. Looks good otherwise

File: code/datums/components/parry.dm [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	var/stamina_constant
+	/// stamina_coefficient * damage * time_since_time_parried = stamina damage taken per non perfect parry
+	var/stamina_coefficient
+	/// the attack types that are considered for parrying
+	var/parryable_attack_types
+	/// the time between parry attempts
+	var/parry_cooldown
+
+/datum/component/parry/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/equipped)
```
Comment (farie82) [inline code/datums/components/parry.dm:17]: Whoops missed these

**Review by farie82 — CHANGES_REQUESTED** [review]

**Comment by hal9000PR** [comment]
UPDATE TWO:
Decreased the parry window further, 1 Second now. increased the cooldown based on the OLD BLOCKCHANCE OF THE ITEM.
if a weapon had a 75% block chance, it has a maximum parry uptime of 75%, 50% blockchance? 50% uptime. etc. 

Training wheels are off.

**Review by farie82 — APPROVED** [review]

**Pull request merged**